ddtrace/tracer/spancontext.go

```diff
@@ -326,11 +326,30 @@ func (t *trace) finishedOne(s *span) {
 	atomic.AddInt64(&tr.spansFinished, int64(len(t.spans)))
 	sd := samplingDecision(atomic.LoadInt64((*int64)(&t.samplingDecision)))
 	if sd != decisionKeep {
+		// if trace sampling decision is drop, we still want to send single spans
+		// unless there are no single span sampling rules defined
+		var singleSpans []*span
+		if tr.rulesSampling.HasSpanRules() {
+			canDropP0s := tr.config.canDropP0s()
+			for _, span := range t.spans {
+				if tr.rulesSampling.sampleSpan(span) {
+					// since stats are computed on the tracer side, the keyTopLevel tag
+					// must be removed to preserve stats correctness
```

Comment:
I don't understand this part. Do you mean `since p0 traces are dropped in the tracer` ? Please note that based on reading this code that is not equal to stats being computed in the tracer.

This does also not explain why we are removing this metric. Can you please clarify?

Reply:
This logic is probably coming from this section of the RFC: https://github.com/DataDog/architecture/blob/master/rfcs/apm/integrations/single-span-ingestion-control/README.md#when-tracers-dont-send-traces-to-the-agent

Some tracers, like the C++ tracer, always send all spans to the agent using the v0.4 API, and so none of this is relevant there.

Other tracers, like the Python tracer, do not send dropped traces to the agent.  Instead, they keep track of aggregated `trace.*` metrics internally and periodically hit a `/stats` endpoint.

The trouble is, how are single-span-sampled spans supposed to get to intake if the tracer is not sending the trace to the agent?  The answer is to send the sampled spans separately, and to make sure that the agent doesn't behave as if it's being sent a sampled trace.

I was not certain how to do this.  The RFC's recipe is the result of some conversations I had, probably with Barbayar and/or Raphaël.

Now that I'm looking at how "trace stats computation" works in practice, though (https://github.com/DataDog/dd-trace-py/pull/2915) I'm wondering why exactly I focused on the "toplevel" concept. Perhaps that part is to prevent the agent from updating priority trace sampling rates based on spans sent to it on account of span sampling rules... but I don't think that's it.  Notice how the RFC says "do these things" and then says "for these reasons," but it doesn't say which thing corresponds to which reason.  It's possible that that section of the RFC is not as it should be. Maybe I thought "the agent calculates stats for toplevel spans -- let's make sure there are no toplevel spans."  That's unnecessary, though, because there is a `Datadog-Client-Computed-Stats: yes` header.

Perhaps the omission of that tag is not necessary after all.

@Kyle-Verhoog @Barbayar @dianashevchenko What do you think?